### PR TITLE
Reflection equality improvements

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/equality/reflection.kt [hunk 1]
```diff
@@ -139,7 +141,6 @@ fun <T : Any> beEqualToUsingFields(other: T, vararg fields: KProperty<*>): Match
  * Note: Throws [IllegalArgumentException] in case [properties] parameter is not provided.
  */
 fun <T : Any> T.shouldBeEqualToIgnoringFields(other: T, vararg properties: KProperty<*>) {
```
Comment (sksamuel) [inline kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/equality/reflection.kt:141]: Do you think this method should allow specifying no fields? Equal to ignoring nothing? That would seem to be a strange way to use this method.

Comment (sschuberth) [reply kotest-assertions/kotest-assertions-core/src/jvmMain/kotlin/io/kotest/matchers/equality/reflection.kt:141]: Well, that's what has been [recommend](https://github.com/kotest/kotest/issues/1161#issuecomment-575724787) to me for me use case. From a strictly logical perspective this sounds alright, but I agree it feels a bit unnatural.

But if neither `shouldBeEqualToUsingFields` nor `shouldBeEqualToIgnoringFields` offer an easy way to compare *all* fields, and preferably also with an option to compare `private` fields, what's the alternative?

**Comment by sksamuel** [comment]
What about adding shouldBeEqualToUsingAllFields ?

**Comment by sksamuel** [comment]
Or if we have to allow empty fields on one of them, I think
shouldBeEqualToUsingFields
makes more sense than the ignored one.

**Comment by sschuberth** [comment]
> I think shouldBeEqualToUsingFields makes more sense than the ignored one.

Yes, that's how I [expected it to work](https://github.com/kotest/kotest/issues/1161#issuecomment-575675811) in the first place. I'll make that change.

**Comment by sksamuel** [comment]
lgtm

**Pull request merged**